core/lib/src/rocket.rs

```diff
@@ -765,19 +781,58 @@ impl Rocket {
             }
         });
 
+        #[cfg(feature = "ctrl_c_shutdown")]
+        let (cancel_ctrl_c_listener_sender, cancel_ctrl_c_listener_receiver) = oneshot::channel();
+
         // NB: executor must be passed manually here, see hyperium/hyper#1537
-        let server = hyper::Server::builder(incoming)
+        let (future, handle) = hyper::Server::builder(incoming)
             .executor(runtime.executor())
-            .serve(service);
+            .serve(service)
+            .with_graceful_shutdown(async move { shutdown_receiver.next().await; })
+            .inspect(|_| {
+                #[cfg(feature = "ctrl_c_shutdown")]
+                let _ = cancel_ctrl_c_listener_sender.send(());
+            })
+            .remote_handle();
 
-        let (future, handle) = server.remote_handle();
         runtime.spawn(future);
-        Ok(handle.err_into())
+
+        #[cfg(feature = "ctrl_c_shutdown")]
+        match tokio::net::signal::ctrl_c() {
+            Ok(mut ctrl_c) => {
+                runtime.spawn(async move {
+                    // Race the futures against each other. In doing so, we
```

Comment:
This is a bit long-winded explanation. Can we simplify this to "Stop listening for Ctrl-C if the server shuts down externally to avoid waiting forever?"